shared/components/data/store-connection/index.jsx

```diff
@@ -27,34 +29,47 @@ var StoreConnection = React.createClass( {
 		}
 	},
 
-	componentWillUnmount: function() {
+	componentWillUnmount() {
 		this.removeStoreListeners( this.props.stores );
 	},
 
-	addStoreListeners: function( stores ) {
+	addStoreListeners( stores ) {
 		stores.forEach( function( store ) {
 			store.on( 'change', this.handleStoresChanged );
 		}, this );
 	},
 
-	removeStoreListeners: function( stores ) {
+	removeStoreListeners( stores ) {
 		stores.forEach( function( store ) {
 			store.off( 'change', this.handleStoresChanged );
 		}, this );
 	},
 
-	handleStoresChanged: function() {
+	handleStoresChanged() {
 		this.setState( this.props.getStateFromStores( this.props ) );
 	},
 
-	render: function() {
+	isDataLoading() {
+		if ( ! this.props.loadingPlaceholder || ! this.props.isDataLoading ) {
+			return false;
+		}
+
+		return this.props.isDataLoading( this.state );
+	},
+
+	render() {
+		if ( this.isDataLoading() ) {
+			return React.createElement( this.props.loadingPlaceholder );
```

Comment:
I see several benefits:
- there is only once place where you do check if data has loaded - it's now in `<StoreConnection />` component, you don't need to repeat it in all page components
- we can put `isDataLoading` method in data component, as we share them between page components we also limit number of imports
- waiting until data is loaded often breaks flow in page component, `<EditCardDetails />` is perfect example, where we had to code some conditions to make sure form gets initialised properly
- with loading logic moved away from page component we can also make assumption that all data is there and therefore make all data `React.PropType` entries mandatory.

This comes at the cost that we need to duplicate some HTML snippets. I'd love to find some solution to that.